metrics/middleware.go

```diff
@@ -0,0 +1,49 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package metrics
+
+import (
+	"net/http"
+
+	"github.com/gorilla/mux"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+// MetricsMiddleware is a middleware used to measure every request received
+func MetricsMiddleware() mux.MiddlewareFunc {
+	// Rergister all metrics
+	prometheus.MustRegister(ServiceInfo)
+
+	prometheus.MustRegister(httpInFlightRequests)
+	prometheus.MustRegister(httpRequestsTotal)
+	prometheus.MustRegister(httpRequestDurationSeconds)
+	prometheus.MustRegister(httpRequestSizeBytes)
+	prometheus.MustRegister(httpResponseSizeBytes)
+
+	prometheus.MustRegister(NumberIndexedPackages)
+	prometheus.MustRegister(CursorUpdatesTotal)
+	prometheus.MustRegister(StorageIndexerGetDurationSeconds)
+	prometheus.MustRegister(StorageIndexerUpdateIndexDurationSeconds)
+	prometheus.MustRegister(StorageIndexerUpdateIndexErrorsTotal)
+	prometheus.MustRegister(StorageRequestsTotal)
+
+	return func(next http.Handler) http.Handler {
+		handler := next
+
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			route := mux.CurrentRoute(req)
+			path, _ := route.GetPathTemplate()
+			labels := prometheus.Labels{"path": path}
```

Comment:
Should we handle the error of `route.GetPathTemplate()`, and set the `path` label only when available?
```suggestion
			path, err := route.GetPathTemplate()
			if err == nil {
				labels := prometheus.Labels{"path": path}
			}
```

Reply:
Good catch, I missed handling this error.

As the prometheus.Labels var must have all the values defined in the metrics (path should have some value), I think I'll go with this:
 
```
 			path, err := route.GetPathTemplate()
			if err != nil {
				path := "unknown"
			}
			labels := prometheus.Labels{"path": path}
```

When I try to query urls that are not defined like `https://localhost:8080/notfound`, the `router.NotFoundHandler` is executed and there is no metric about that in metrics endpoint. That's why I think that error did not raise in my tests.